kong/utils.go

```diff
@@ -741,3 +741,147 @@ func FillPluginsDefaults(plugin *Plugin, schema Schema) error {
 	return fillConfigRecordDefaultsAutoFields(plugin, schema, opts)
 }
+
+// deleteAndCollapseMap is a utility function that removes element from a map using path.
+// If it turns out that that was the last key in the map then it collapses map key itself.
+// Example:
+//
+//		configMap := map[string]interface{}{
+//			"a": map[string]interface{}{
+//				"a1": 1,
+//			},
+//	   "b": 456,
+//		}
+//
+// then by calling deleteAndCollapseMap(configMap, []string{"a", "a1"})
+// // the result would be: (entire "a" key is deleted as well)
+//
+//	map[string]interface{} {
+//	  "b": 456,
+//	}
+func deleteAndCollapseMap(config map[string]interface{}, path []string) {
+	if len(path) == 0 {
+		return
+	}
+
+	key := path[0]
+	if len(path) == 1 {
+		delete(config, key)
+		return
+	}
+
+	if nested, ok := config[key].(map[string]interface{}); ok {
+		deleteAndCollapseMap(nested, path[1:])
+		if len(nested) == 0 {
+			delete(config, key)
+		}
+	}
+}
+
+// same as ClearUnmatchingDeprecations but this function below is adjusted for recursive use (which is required
+// when the schema contains nested records and they can have nested shorthands).
+func clearUnmatchingDeprecationsHelper(
+	newPluginConfig Configuration,
+	oldPluginConfig Configuration,
+	schema *gjson.Result,
+) {
+	configFields := schema.Get("fields")
+	// Fetch deprecated fields
+	shortHandFields := schema.Get("shorthand_fields")
+
+	shortHandFields.ForEach(func(_, value gjson.Result) bool {
+		field := value.Map()
+		for deprecatedFieldName, shorthandFieldConfig := range field {
+			if deprecatedFieldValue, ok := newPluginConfig[deprecatedFieldName]; ok {
+				// deprecatedFieldName is used in new plugin configuration
+				// verify if the fields that this depractedField is replaced with
+				// are also sent in new plugin configuration - if not clear them from old plugin configuration
+				replacements := shorthandFieldConfig.Get("deprecation.replaced_with.#.path")
+				replacements.ForEach(func(_, value gjson.Result) bool {
+					replacementPathAsStringArray := make([]string, len(value.Array()))
+					for i, pathSegment := range value.Array() {
+						replacementPathAsStringArray[i] = pathSegment.String()
+					}
+
+					// We know that deprecated value is defined in new config and we also have information
+					// on how this deprecated value is replaced. If the new plugin configuration contains
+					// both old and new values we don't need to adjust old plugin configuration.
+					// However if the new key is missing in new plugin configuration then we need to
+					// delete it from old plugin configuration in order for them to match.
+					// --
+					// There is also a possibility that the new field exists in new config but only because
+					// it was filled with `null` value due to decK logic of filling empty keys. In that case
+					// if the deprecated value is different from nil then we need to clear it both in new and old config.
```

Comment:
I think a bullet-list with all the different scenarios, and what we do for each, would be more helpful as a description of the code below

Reply:
Fixed ✅